client/src/components/Visualiser/VisualiserCanvas.tsx

```diff
@@ -1,116 +1,215 @@
-    setNewViewBox({
-      x: Math.min(width, Math.max(-width, viewBox.x - (event.clientX - pointerOrigin.x))),
-      y: Math.min(height, Math.max(-height, viewBox.y - (event.clientY - pointerOrigin.y))),
-    });
-  };
+  const handlePointerMove = useCallback(
+    (event: PointerEvent<HTMLDivElement>) => {
+      if (!isPointerDown) {
+        return;
+      }
+
+      // If the mouse is outside the workspace, don't do translation
+      if (
+        workspaceRect &&
+        (event.clientX < workspaceRect.left ||
+          event.clientY < workspaceRect.top ||
+          event.clientX > workspaceRect.right ||
+          event.clientY > workspaceRect.bottom)
+      ) {
+        if (DEBUG) console.log('Pointer left workspace');
+        setIsPointerDown(false);
+        return;
+      }
+
+      // Is this needed for anything?
+      // event.preventDefault();
+
+      setTransform((prev) => {
+        const translateVec = vec2.fromValues(event.movementX / prev[0], event.movementY / prev[4]);
+        return mat3.translate(mat3.create(), prev, translateVec);
+      });
+    },
+    [isPointerDown, workspaceRect]
+  );
 
   const handlePointerUp = () => {
     setIsPointerDown(false);
-
-    setViewBox({
-      x: newViewBox.x,
-      y: newViewBox.y,
-    });
   };
 
   useEffect(() => {
-    setHeight(svgRef.current.clientHeight);
-    setWidth(svgRef.current.clientWidth);
-    setViewBox((prevViewBox) => ({
-      ...prevViewBox,
-      height,
-      width,
-    }));
+    // Callback for a ResizeObserver constructor
+    const handleWorkspaceResize: ResizeObserverCallback = (entries) => {
+      if (entries.length > 0 && entries[0].target.id === VISUALISER_WORKSPACE_ID) {
+        const workspaceEntry = entries[0];
+        const boundingClientRect = workspaceEntry.target.getBoundingClientRect();
+        if (DEBUG)
+          console.log(
+            `Workspace resized. Setting new workspace rect left: ${boundingClientRect.left}, right: ${boundingClientRect.right}, top: ${boundingClientRect.top}, bottom: ${boundingClientRect.bottom} and updating transform...`
+          );
+        setWorkspaceRect(boundingClientRect);
+      }
+
+      if (entries.length !== 1) {
+        // Note: I don't know why there would be none or multiple entries, I just assume
+        // there is one entry passed to the resize handler when it is called by
+        // the resize observer. So if this is not the case, maybe investigate\
+        // what to do.
+        console.warn(
+          'Warning: Unexpected number of entries in ResizeObserver callback (more than 1).'
+        );
+      }
+    };
+
+    const visualiserWorkspaceEle = document.getElementById(VISUALISER_WORKSPACE_ID);
+    if (visualiserWorkspaceEle) {
+      // https://developer.mozilla.org/en-US/docs/Web/API/Resize_Observer_API
+      new ResizeObserver(handleWorkspaceResize).observe(visualiserWorkspaceEle);
+    }
   }, []);
```

Comment:
Use `ResizeObserver` to listen to workspace resize events and keep `workspaceRect` state up-to-date, which keeps user pan/zoom interacts consistent with workspace size.